File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        l, r = 0, 0
        profit = 0
        
        while r < len(prices):
            diff = prices[r] - prices[l]
            profit = max(profit, diff)
            if prices[r] < prices[l]:
                l = r
            r += 1
        
        return profit

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int min_val = Integer.MAX_VALUE;
        int max_val = 0;
        
        for(int i = 0; i < prices.length; i++){
            if(prices[i] < min_val){
                min_val = prices[i];
            }
            else if(max_val < prices[i] - min_val){
                max_val = prices[i] - min_val;
            }
        }
        return max_val;
    }
}